### PR TITLE
Implement HttpRequestError

### LINKED ISSUE #76644: [API Proposal]: Support error codes for HttpRequestException
*Edited by @antonfirsov*
*Update 1: changes based on feedback - https://github.com/dotnet/runtime/issues/76644#issuecomment-1525754867*

### Background and motivation

HttpRequestException is thrown under various conditions but it does not provide an error code to differentiate those conditions.
Currently, it provides 'string Message' property inherited from Exception type and it can contain various error messages:
- The HTTP response headers length exceeded the set limit
- Cannot write more bytes to the buffer
- The response ended prematurely
- Received chunk header length could not be parsed
- Received an invalid status
- Received an invalid header line
- Received an invalid header name
- The server returned an invalid or unrecognized response
- etc.

The above strings are defined in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/Resources/Strings.resx

If we want to pragmatically handle exceptions based on the cause, it would be better dealing with error codes rather than dealing with string values. For example, when we create a dashboard for different exceptions encountered, aggregating by error codes would make more sense that doing by string values. And we could categorize some errors into a group for a special handling.

Many other exceptions like SocketException and Win32Exception have an error code field already.
Would it make sense to provide a new property for HttpRequestException to hold error codes?

### API Proposal

```C#

public class HttpRequestException : Exception
{
    // -- New properties --
    public HttpRequestError HttpRequestError { get; }

    // -- New constructors --
    public HttpRequestException(HttpRequestError httpRequestError, string? message) {}
    public HttpRequestException(HttpRequestError httpRequestError, string? message, Exception? inner) {}
    public HttpRequestException(HttpRequestError httpRequestError, string? message, Exception? inner, HttpStatusCode? statusCode) {}
    // ALTERNATIVE:
    public HttpRequestException(HttpRequestError httpRequestError, string? message, Exception? inner = null, HttpStatusCode? statusCode = null) {}
}

// IOException subtype to throw from response read streams
public class HttpIOException : IOException  // ALTERNATIVE name: HttpResponseReadException
{
    public HttpRequestError HttpRequestError { get; }
    public HttpResponseReadException(HttpRequestError httpRequestError, string? message = null, Exception? innerException = null) {}
} 

// The EXISTING HttpProtocolException should be the subtype of this new exception type
public class HttpProtocolException : HttpResponseReadException // WAS: IOException
{
}

// Defines disjoint error categories with high granularity.
public enum HttpRequestError
{
    Undefined = 0,                          // Uncategorized/generic error -OR-
                                            // the underlying HttpMessageHandler does not implement HttpRequestError

    NameResolutionError,                    // DNS request failed
    ConnectionError,                        // Transport-level error during connection
    TransportError,                         // Transport-level error after connection
    SecureConnectionError,                  // SSL/TLS error
    HttpProtocolError,                      // HTTP 2.0/3.0 protocol error occurred
    UnsupportedExtendedConnect,             // Extended CONNECT for WebSockets over HTTP/2 is not supported.
                                            // (SETTINGS_ENABLE_CONNECT_PROTOCOL has not been sent).
    VersionNegotiationError,                // Cannot negotiate the HTTP Version requested
    UserAuthenticationError,                // Authentication failed with the provided credentials
    ProxyTunnelError,

    StatusNotSuccess,                       // Set by EnsureSuccessStatusCode() in case of non-success status code.
    InvalidResponse,                        // General error in response/malformed response
    
    // ALTERNATIVE name: InvalidResponseHeaders
    InvalidResponseHeader,                  // Error with response headers
    
    // OPTIONAL TO NOT INCLUDE: The one's below are specific cases of errors with the response.
    // We may consider merging them into "InvalidResponse" and "InvalidResponseHeader"
    ResponseEnded,                          // Received EOF
    // These depend on SocketsHttpHandler configuration
    ContentBufferSizeExceeded,              // Response Content size exceeded MaxResponseContentBufferSize
    ResponseHeaderExceededLengthLimit,      // Response Header length exceeded MaxResponseHeadersLength
}
```

### API Usage

With `HttpClient` methods:

```C#
try
{
    using var response = await httpClient.GetStringAsync(url);
}
catch (HttpRequestException ex) when (ex.HttpRequestError is HttpRequestError.ConnectionError or
                                                             HttpRequestError.SecureConnectionError or
                                                             HttpRequestError.NameResolutionError)
{
    // Retry
}
```

With response read `Stream`:

```C#
using var response = await httpClient.GetAsync(url, HttpCompletionOption.ResponseHeadersRead);
using var responseStream = await response.Content.ReadAsStreamAsync();
try
{
    await responseStream.ReadAsync(buffer);
}
catch (HttpResponseReadException ex)
{
    switch (ex.HttpRequestError)
    {
        // ....
    }
}
```

### Risks
Any change to the error categorization is a breaking change, meaning that if we introduce new error codes for some users, we would break others.

Mobile handlers would need to implement their error mapping separately, meaning that we would lack platform consistency until that point. Whenever it's impossible to detect some error categories on those platforms, they would need to report a more generic error code eg. `InvalidResponse` instead of `ResponseEnded`. This is a challenge for users who need to include mobile targets in their cross-platform `HttpClient` code.

Since requests and connections are decoupled, connection-establishment errors (`NameResolutionError`, `ConnectionError` and `SecureConnectionError`) will be only reported if the originating request is still in the queue, otherwise we will swallow them. Observing these error codes is not a reliable way to observe ALL connection-level errors happening under the hood. Some users may find this surprising.

### Alternative Designs

- Move errors that can occur during response content read to a separate `HttpResponseReadError` enum. `HttpResponseReadException` would hold values from that separate enum.
    - Upside: This might make it easier for users to implement unified error handling for the two API scenarios (`HttpClient` and `Stream`).
    - Downsides: might be harder to follow. Some error cases like `TransportError` or `ProtocolError` can occur in bothe cases, meaning that enum values would be shared.
- Reduce the granularity of the error codes, eg. avoid defining cases like `ContentBufferSizeExceeded`, and threat them as `InvalidResponse`.
    - Upside: Easier to implement and maintain.
    - Downside: we won't be able to handle valid error scenarios like the one mentioned in https://github.com/dotnet/runtime/issues/76644#issuecomment-1367521636
- Use exception subclasses instead of error codes for better extensibility. The most practical approach would be to keep `HttpRequestException` sealed, and define a class hierarchy under `IOException` (which then should be aways `InnerException` of `HttpRequestException`).
    - Downside: too many subclasses, such hierarchy is unprecedented in the BCL

### LINKED ISSUE #82168: HttpProtocolException is not reported properly on initial HTTP/2 settings handshake failure
There is an overlook in #71345. When the server doesn't initiate the connection with a proper SETTINGS preface, eg. responding by GOAWAY instead, the `InnerException` of the `HttpRequestException` being thrown must be a `ProtocolException` according to #70684.

Instead, we embed `ProtocolException` into further `IOException`-s  2 levels down in [`ProcessIncomingFramesAsync()`](https://github.com/dotnet/runtime/blob/0202b242169f8d5955e0411dc5e4f42a5d016c1b/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L503).

Repro:

```C#
[ConditionalFact(nameof(SupportsAlpn))]
public async Task Test()
{
    using Http2LoopbackServer server = Http2LoopbackServer.CreateServer();
    using HttpClient client = CreateHttpClient();
            
    Task<HttpResponseMessage> sendTask = client.GetAsync(server.Address);

    Http2LoopbackConnection connection = await server.AcceptConnectionAsync();
    await connection.ReadSettingsAsync();
    await connection.SendGoAway(0, ProtocolErrors.INTERNAL_ERROR);

    var ex = await Assert.ThrowsAsync<HttpRequestException>(() => sendTask);
    Assert.IsType<HttpProtocolException>(ex.InnerException); // Fails
}
```

### PR DESCRIPTION
**Pull request by antonfirsov**
Implements the proposal from https://github.com/dotnet/runtime/issues/76644#issuecomment-1624144878. I left out `TransportError`, since it would be a new error case with little value, requiring new error logic.

Fixes #76644, fixes #82168.

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (antonfirsov) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:259]: When can these cases occur? How can we test them? Can we return something better for `HttpRequestError.Unknown`? Same for the other cases where I'm using `HttpRequestError.Unknown`.

File: src/libraries/System.Net.Http/ref/System.Net.Http.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/ref/System.Net.Http.cs:1]: These changes look different from what the tool would generate (e.g. fully spelled-out types)

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:536]: Allocating separate exception instances here was intentional: https://github.com/dotnet/runtime/pull/80066#discussion_r1059515850

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:1162]:
```suggestion
```

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:1234]:
```suggestion
```

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:96]: The existence of these data entries is unfortunately a part of the contract at this point since we shipped it in .NET 7. I don't think we can remove them (at least not in Net 8).

[YARP relies on that](https://github.com/microsoft/reverse-proxy/blob/f74196b1f643e422f5f95f6aab6e39dde66eabfc/src/ReverseProxy/Forwarder/HttpForwarder.cs#L166-L180) in its HTTP/2 WebSocket implementation. Removing them would break the automatic version downgrade if users started running on .NET 8.

(same for the other 2 places that modified these)

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs:362]: How would errors during the establishment of a CONNECT proxy tunnel be surfaced (e.g. proxy sent a garbled response -- not just a valid but non-200 one)?

`ProxyTunnelError`'s description also seems to fit this case (we got a connection but failed to agree with the proxy).

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpIOException.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpIOException.cs:9]:
```suggestion
    /// An exception thrown when an error occurs while reading the response.
```

Nit: "content stream" reads to me like it's referring only to the body

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:259]: The way I understand it, these do not surface to the user, they mean that we should retry on a different connection (either different version, or just a brand new H/3 connection). Looking at the handling code (after some retries, they might bubble up though).
The first one gets wrapped:
https://github.com/dotnet/runtime/blob/b898a540f75d78ca55ee33288a4fd62b1ab69fa2/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1165
The second won't:
https://github.com/dotnet/runtime/blob/b898a540f75d78ca55ee33288a4fd62b1ab69fa2/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1142

But, it's the same thing as this:
https://github.com/dotnet/runtime/blob/b898a540f75d78ca55ee33288a4fd62b1ab69fa2/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L2205-L2206

**which you don't touch in this PR, so do we need to touch these at all????**

Also, you should be easily able to test it with our loopback set up:
- RequestRejected: https://github.com/dotnet/runtime/blob/d4db2d96bb0f55211c30dc69ac587b204bb418f3/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs#L384-L384
- VersionFallback: seems like we don't have a test for it, or at least I cannot find anything.

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestException.cs
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestException.cs:57]: Why is this nullable if we have Unknown, or the other way around?

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestException.cs:57]: > and then agreeing with the advice from the API review group that nullable actually makes sense.

I don't understand why nullable makes sense here. I was away for the API review, and see the comments in the issue:

> Changed the HttpRequestError property on HttpRequestException to be nullable, to distinguish "the provider didn't give one" from "the provider had an error it couldn't map".

but I agree that this is a confusing distinction to try to make.  What action is someone supposed to take differently based on null vs Unknown?

cc: @bartonjs, @terrajobst

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpIOException.cs
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpIOException.cs:16]: Nit: normally we put properties below constructors

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpIOException.cs
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpIOException.cs:28]: The HttpRequestError isn't going to be included automatically in either the Message or ToString for this exception.  Do you want to override?

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestException.cs
Comment (bartonjs) [reply src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestException.cs:57]: The main distinction that I can think of would be "I should try again if the error is [some specific code(s)], and not if not, and Unknown is clearly not one of those codes".  If it's null (vs Unknown) then the caller can decide "the underlying provider doesn't know how to set this code" is different from "something went wrong that the caller didn't know how to map to this enum"

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestException.cs:57]: I still don't understand the distinction / how someone would apply it.  You're saying someone would write code that would retry on null but not on Unknown, or retry on Unknown but not null?  Whether the provider knows what the problem is but can't map it, or doesn't know and can't map it, the result is the same: the error is unknown and you as the consumer have zero additional information to inform you as to what to do with it.  I'm not understanding why we'd choose two different representations for that same case.

File: src/libraries/Common/tests/System/Net/Http/ResponseStreamTest.cs
Comment (wfurt) [inline src/libraries/Common/tests/System/Net/Http/ResponseStreamTest.cs:279]: we probably should remove this. Or show it only on test failure

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs]: Added a new message for the topmost `HttpRequestException` and decided to leave inner SocksExceptions as they are.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs]: I handled these cases in 9afb7900d89ec6421225ea84c3b9f8185f1843b9, by using `HttpRequestError.ProtocolError` and `ProtocolException` as an inner exception, but there are some other HTTP/3 cases which are still use `Unknown`:

https://github.com/dotnet/runtime/blob/2eb3a4a8e136aeca554372d10f20d514898a56a4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs#L277-L281

https://github.com/dotnet/runtime/blob/2eb3a4a8e136aeca554372d10f20d514898a56a4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs#L291-L295

https://github.com/dotnet/runtime/blob/2eb3a4a8e136aeca554372d10f20d514898a56a4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs#L302-L309

I don't know what can trigger those, and whether there is a better `HttpRequestError` we could use. /cc @wfurt

File: src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs
Comment (antonfirsov) [inline src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:4374]: I believe we should not block the merge on this, I plan to open a separate issue for the investigation.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (wfurt) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:146]: 87262 is merged. can be updated later.

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestError.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestError.cs:52]:
```suggestion
        /// An error occurred while establishing a proxy tunnel.
```
If this is used more broadly than just when connecting to the proxy.
Same for the `Strings.resx` entry.

Nit: Find-replace `occured` => `occurred`

**Review by wfurt — APPROVED** [review]
looks good enough for me. I think we can resolve outstanding issues as follow-up

**Comment by wfurt** [comment]
it seems ike test failures may be related - for other platforms. We should ideally resolve before merging @antonfirsov. (or at least disable so we do not pollute CI)

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:2026]: Is `UserAuthenticationError` meant to encapsulate only "wrong username/password" or "something went wrong with auth"?
The description sounded to me like the former, but this looks like the latter.

(same for the exception below)

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1104]: This should stay (https://github.com/dotnet/runtime/pull/88974#discussion_r1264933744).

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs:138]:
```suggestion
                        ((ex.HttpRequestError is  HttpRequestError.ExtendedConnectNotSupported or HttpRequestError.VersionNegotiationError)
```
We should be able to rely on the error code but keep the `HTTP2_ENABLED` in the code for backward compat.

**Review by MihaZupan — APPROVED** [review]
Left a few comments, but `SETTINGS_ENABLE_CONNECT_PROTOCOL` is the only one that should be resolved before merging.

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs
Comment (antonfirsov) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs:138]: Shouldn't it also include `SecureConnectionError`
```suggestion
                        ((ex.HttpRequestError is HttpRequestError.ExtendedConnectNotSupported or HttpRequestError.VersionNegotiationError or HttpRequestError.SecureConnectionError)
```

given that this what we detect in `ConnectHelper.EstablishSslConnectionAsync`:

https://github.com/dotnet/runtime/blob/633d632fb5178245964b41d485cf6f64683657c1/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs#L92C18-L98

Or would that also catch TLS errors we don't want to catch here?

Comment (MihaZupan) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs:138]: Looks like there's an overlap with #75399.

If you specify `Version: 2.0, Policy: Exact` and the server doesn't support HTTP/2, you will get a `SecureConnectionError` error, not a `VersionNegotiationError` with this change.

For H2 WebSockets, we would want to handle that case, but only if HTTP/2 was the problem, not for any TLS error.

I guess we can keep it as-is for now (checking `HTTP2_ENABLED`) and then decide if we can resolve #75399 in a way that makes using the data unnecessary.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:2026]: Update the docs for `UserAuthenticationError`.

**Pull request merged**
